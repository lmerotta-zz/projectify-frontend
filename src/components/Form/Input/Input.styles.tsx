import tw, { styled } from "twin.macro";
import { motion } from "framer-motion";

export const Wrapper = motion(
  tw.div`outline-none w-full flex flex-col items-stretch`
);

export const InputWrapper = motion(tw.div`w-full relative`);
export const Input = styled.input<{ invalid?: boolean }>`
  ${tw`w-full p-2 text-base border border-gray-400 bg-white appearance-none outline-none transition-colors focus:border-primary`}

  ${({ invalid }) => invalid && tw`border-danger!`}
`;
