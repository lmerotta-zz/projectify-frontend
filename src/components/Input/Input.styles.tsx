import tw, { styled } from "twin.macro";
import { motion } from "framer-motion";

export const Wrapper = motion(tw.div`relative outline-none`);

export const labelFilledStyles = tw`text-primary top-0 translate-y-0.5 text-xs`;

export const Label = styled.label`
  ${tw`absolute top-1/2 transform -translate-y-1/2 left-3 text-default cursor-text transition-all pointer-events-none`}
`;

export const InputWrapper = motion(tw.div`w-full relative`);
export const Input = styled.input<{ invalid?: boolean }>`
  ${tw`w-full pt-4 pb-2 pl-3 pr-3 border rounded border-gray-400 appearance-none outline-none transition-colors focus:border-primary`}
  &:focus {
    + .label {
      ${labelFilledStyles}
    }
  }

  :not(:placeholder-shown) {
    ${tw`border-primary`}
    + .label {
      ${labelFilledStyles}
    }
  }

  ${({ invalid }) => invalid && tw`border-red-600!`}
  + .label {
    ${({ invalid }) => invalid && tw`text-red-600!`}
  }
`;
