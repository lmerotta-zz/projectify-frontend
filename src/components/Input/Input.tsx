import { forwardRef, HTMLProps, ReactNode } from "react";
import tw, { styled } from "twin.macro";
import { AnimateSharedLayout, motion } from "framer-motion";
import ErrorMessage from "components/ErrorMessage/ErrorMessage";

const Wrapper = motion.custom(tw.div`relative outline-none`);

const labelFilledStyles = tw`text-primary top-0 translate-y-0.5 text-xs`;

const Label = styled.label`
  ${tw`absolute top-1/2 transform -translate-y-1/2 left-3 text-default cursor-text transition-all pointer-events-none`}
`;

const InputWrapper = motion.custom(tw.div`w-full relative`);
const StyledInput = styled.input`
  ${tw`w-full pt-4 pb-2 pl-3 pr-3 border rounded border-gray-400 appearance-none outline-none transition-colors focus:border-primary`}
  &:focus {
    + .label {
      ${labelFilledStyles}
    }
  }

  :not(:placeholder-shown) {
    ${tw`border-primary`}
    + .label {
      ${labelFilledStyles}
    }
  }

  &:invalid {
    ${tw`border-red-600`}
    + .label {
      ${tw`text-red-600`}
    }
  }
`;

type InputProps = Omit<HTMLProps<HTMLInputElement>, "as" | "name"> & {
  label: ReactNode;
  name: string;
};

const Input = forwardRef<HTMLInputElement, InputProps>(
  ({ label, name, ...rest }, ref) => (
    <AnimateSharedLayout>
      <Wrapper layout>
        <InputWrapper layout>
          <StyledInput
            ref={ref}
            {...rest}
            name={name}
            placeholder="&nbsp;"
            data-testid={`input-${name}`}
          />
          {label && (
            <Label className="label" data-testid={`label-${name}`}>
              {label}
            </Label>
          )}
        </InputWrapper>
        <ErrorMessage name={name} />
      </Wrapper>
    </AnimateSharedLayout>
  )
);

export default Input;
