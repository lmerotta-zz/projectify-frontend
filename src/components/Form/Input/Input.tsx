import { forwardRef, HTMLProps, ReactNode } from "react";
import * as Styles from "./Input.styles";
import ErrorMessage from "../ErrorMessage/ErrorMessage";
import { useFormContext } from "react-hook-form";

type InputProps = Omit<HTMLProps<HTMLInputElement>, "as" | "name"> & {
  label: ReactNode;
  name: string;
};

const Input = forwardRef<HTMLInputElement, InputProps>(
  ({ label, name, ...rest }, ref) => {
    const { errors } = useFormContext();
    return (
      <Styles.Wrapper>
        <Styles.InputWrapper layout>
          <Styles.Input
            ref={ref}
            {...rest}
            name={name}
            placeholder="&nbsp;"
            data-testid={`input-${name}`}
            invalid={errors[name]?.length > 0}
          />
          {label && (
            <Styles.Label className="label" data-testid={`label-${name}`}>
              {label}
            </Styles.Label>
          )}
        </Styles.InputWrapper>
        <ErrorMessage name={name} />
      </Styles.Wrapper>
    );
  }
);

export default Input;
