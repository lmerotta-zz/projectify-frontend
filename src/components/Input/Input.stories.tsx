// Button.stories.tsx
import { ComponentProps, useEffect, useRef } from "react";
import Input from "./Input";
import { Story } from "@storybook/react/types-6-0";
import { FormProvider, useForm } from "react-hook-form";

// eslint-disable-next-line import/no-anonymous-default-export
export default {
  title: "Input",
  component: Input,
  argTypes: {
    label: { type: "string", defaultValue: "Label" },
    name: { defaultValue: "input-test" },
    hasError: { type: "boolean" },
  },
};

const Template: Story<ComponentProps<typeof Input> & { hasError?: string }> = ({
  hasError,
  name,
  ...args
}) => {
  const form = useForm();
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (hasError && !form.errors[name]) {
      form.setError(name, { message: "error message" });
      inputRef.current?.setCustomValidity("blep");
    }
  }, [hasError, form, name]);

  return (
    <FormProvider {...form}>
      <Input {...args} name={name} ref={inputRef} />
    </FormProvider>
  );
};

export const Default = Template.bind({});
export const WithError = Template.bind({});
WithError.args = {
  // @ts-ignore
  hasError: true,
};
