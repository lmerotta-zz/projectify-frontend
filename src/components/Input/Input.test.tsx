import { render } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import Input from "./Input";

describe("Input unit tests", () => {
  it("Applies focusing styles when input is focused", async () => {
    const Component = () => {
      return (
        <FormProvider {...useForm()}>
          <Input name="my-input" label="My input" />
        </FormProvider>
      );
    };
    const { asFragment } = render(<Component />);

    const firstRender = asFragment();
    expect(firstRender).toMatchSnapshot();
  });
});
