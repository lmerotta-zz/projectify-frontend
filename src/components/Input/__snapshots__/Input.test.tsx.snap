// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input unit tests Applies focusing styles when input is focused 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
  outline: 2px solid transparent;
  outline-offset: 2px;
}

.emotion-1 {
  width: 100%;
  position: relative;
}

.emotion-2 {
  width: 100%;
  padding-top: 1rem;
  padding-bottom: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  border-width: 1px;
  border-radius: 0.25rem;
  --tw-border-opacity: 1;
  border-color: rgba(156, 163, 175, var(--tw-border-opacity));
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  outline: 2px solid transparent;
  outline-offset: 2px;
  transition-property: background-color,border-color,color,fill,stroke;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
}

.emotion-2:focus {
  --tw-border-opacity: 1;
  border-color: rgba(147, 51, 234, var(--tw-border-opacity));
}

.emotion-2:focus+.label {
  --tw-text-opacity: 1;
  color: rgba(147, 51, 234, var(--tw-text-opacity));
  top: 0px;
  --tw-translate-y: 0.125rem;
  font-size: 0.75rem;
  line-height: 1rem;
}

.emotion-2:not(:placeholder-shown) {
  --tw-border-opacity: 1;
  border-color: rgba(147, 51, 234, var(--tw-border-opacity));
}

.emotion-2:not(:placeholder-shown)+.label {
  --tw-text-opacity: 1;
  color: rgba(147, 51, 234, var(--tw-text-opacity));
  top: 0px;
  --tw-translate-y: 0.125rem;
  font-size: 0.75rem;
  line-height: 1rem;
}

.emotion-3 {
  --tw-translate-x: 0;
  --tw-translate-y: -50%;
  --tw-rotate: 0;
  --tw-skew-x: 0;
  --tw-skew-y: 0;
  --tw-scale-x: 1;
  --tw-scale-y: 1;
  -webkit-transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  -moz-transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  -ms-transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  position: absolute;
  top: 50%;
  left: 0.75rem;
  --tw-text-opacity: 1;
  color: rgba(161, 161, 170, var(--tw-text-opacity));
  cursor: text;
  transition-property: all;
  transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);
  transition-duration: 150ms;
  pointer-events: none;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <input
        class="emotion-2"
        data-testid="input-my-input"
        name="my-input"
        placeholder="Â "
      />
      <label
        class="label emotion-3"
        data-testid="label-my-input"
      >
        My input
      </label>
    </div>
  </div>
</DocumentFragment>
`;
